### PR TITLE
Transport shutdown timing fix

### PR DESCRIPTION
**Pull request by was4**
* Deconflict lock flags
 * Debug negative references
 * Transport shutdown timing fix

### CODE REVIEW

**Review by dang — COMMENTED** [review]
Okay, so this doesn't fix the issue.  I have an easy reproducer now (for some definition of easy) and it still happens with this applied.

I'll start instrumenting, and we can discuss tomorrow.

File: ntirpc/rpc/clnt.h [hunk 1]
```diff
@@ -114,7 +114,7 @@ typedef struct rpc_client {
 
 	mutex_t cl_lock;	/* serialize private data */
 	struct rpc_err cl_error; /* specific error code */
-	uint32_t cl_refcnt;	/* handle refcnt */
```
Comment (mattbenjamin) [inline ntirpc/rpc/clnt.h:117]: what does changing refcnt to refs do?  I would request that we stay consistent with other uses of refcnt in the ntirpc and nfs-ganesha source

File: src/clnt_internal.h [hunk 2]
```diff
@@ -46,7 +46,7 @@ static inline void
 clnt_data_init(struct cx_data *cx)
 {
 	mutex_init(&cx->cx_c.cl_lock, NULL);
-	cx->cx_c.cl_refcnt = 1;
```
Comment (mattbenjamin) [inline src/clnt_internal.h:49]: ditto

**Review by mattbenjamin — COMMENTED** [review]
what was being being inserted into an rbtree on a per call basis?  this does not appear to be the reply matchin case?

that said "Client call timeout is most likely increasing relative to other calls
for the same connection, worst case for rbtree insertion. Profiling
showed that rbtree was not a good choice, and TAILQ improved by 200%" seems suspicious.  why do we need to update a timer list on every call?

**Comment by was4** [comment]
On 4/25/18 12:47 PM, Daniel Gryniewicz wrote:
> *@dang* commented on this pull request.
> 
> Okay, so this doesn't fix the issue. I have an easy reproducer now (for some definition of easy) and it still happens with this applied.
> 
> I'll start instrumenting, and we can discuss tomorrow.
> 
Good thing I checked email before my nap.  I'd just pushed an
update patch.

I've found one fairly old issue so far:

-                       rpc_dplx_rui(rec);
+                       if (!(flags & RPC_DPLX_LOCKED))
+                               rpc_dplx_rui(rec);

So in that case (should be rare) it was unilaterally unlocked, even
though it wasn't supposed to be.

Found it because I was resolving some symbol overlap a bit earlier:

-       if (!(flags & SVC_RQST_FLAG_LOCKED))
+       if (!(flags & RPC_DPLX_LOCKED))

****_FLAG_LOCKED should be the xp_lock (0x00010000), not the
wrapping rpc_dplx_rec lock.  I've updated those symbols (0x00100000),
as well as an internal SVC_RQST_LOCKED (0x01000000) for good measure.

I remember fixing that sometime in the past year.  But it seems that
I'd accidentally dropped the patch somehow....

Anyway, I've also found that timeouts are sometimes going backward,
so I've given up on compatibility and changed fields to timespec.

And still have a refcounting problem somewhere.

**Comment by was4** [comment]
Likewise, this was another was16backport.  It fixed my crash
under heavy load.  Has this made it into the tree?

**Pull request closed without merging**